### PR TITLE
KAFKA-16532; Support for first leader bootstrapping the voter set

### PR DESCRIPTION
**Pull request by ahuang98**
The first leader of a KRaft topic partition must rewrite the content of the bootstrap checkpoint (0-0.checkpoint) to the log so that it is replicated. Bootstrap checkpoints are not replicated to the followers.

The control records for KRaftVersionRecord and VotersRecord in the bootstrap checkpoint will be written in one batch along with the LeaderChangeMessage. The leader will write these control records before accepting data records from the state machine (Controller).

The leader determines that the bootstrap checkpoint has not been written to the log if the latest set of voters is located at offset -1. This is the last contained offset for the bootstrap checkpoint (0-0.checkpoint).

This change also improves the RaftClientTestContext to allow for better testing of the reconfiguration functionality. This is mainly done by allowing the voter set to be configured statically or through the bootstrap checkpoint.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 1]
```diff
@@ -151,6 +151,7 @@ public final class KafkaRaftClient<T> implements RaftClient<T> {
     public static final int MAX_FETCH_WAIT_MS = 500;
     public static final int MAX_BATCH_SIZE_BYTES = 8 * 1024 * 1024;
     public static final int MAX_FETCH_SIZE_BYTES = MAX_BATCH_SIZE_BYTES;
+    public static final OffsetAndEpoch BOOTSTRAP_SNAPSHOT_ID = new OffsetAndEpoch(0, 0);
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: Let's define this in `o.a.k.s.Snapshots` instead.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 2]
```diff
@@ -454,6 +455,7 @@ public void initialize(
             nodeId,
             nodeDirectoryId,
             partitionState::lastVoterSet,
+            partitionState::lastVoterSetOffset,
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: It is unfortunate that we need to add a parameter to the constructor just so that it can pass this value to `LeaderState`. I am starting to think that `QuorumState` should just have access to the `partitionState` (`KRaftControlRecordStateMachine`) object.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 3]
```diff
@@ -1396,12 +1398,12 @@ private FetchResponseData tryCompleteFetchRequest(
             int lastFetchedEpoch = request.lastFetchedEpoch();
             LeaderState<T> state = quorum.leaderStateOrThrow();
 
-            Optional<OffsetAndEpoch> latestSnapshotId = log.latestSnapshotId();
+            OffsetAndEpoch latestSnapshotId = log.latestSnapshotId().orElse(null);
             final ValidOffsetAndEpoch validOffsetAndEpoch;
-            if (fetchOffset == 0 && latestSnapshotId.isPresent()) {
-                // If the follower has an empty log and a snapshot exist, it is always more efficient
+            if (fetchOffset == 0 && latestSnapshotId != null && !latestSnapshotId.equals(BOOTSTRAP_SNAPSHOT_ID)) {
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: Let's avoid the use of `null`.
```java
            if (fetchOffset == 0 &&
                latestSnapshotId.isPresent() &&
                !latestSnapshotId.get().equals(BOOTSTRAP_SNAPSHOT_ID)
            ) {
```

Reference: https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare/

File: raft/src/main/java/org/apache/kafka/raft/LeaderState.java [hunk 4]
```diff
@@ -202,7 +214,46 @@ public void appendLeaderChangeMessage(long currentTimeMs) {
...
+                buffer.capacity()
+            )
+            ) {
+                builder.appendLeaderChangeMessage(
+                    currentTimeMs,
+                    leaderChangeMessage
+                );
+                VoterSetOffset voterSetOffset = lastVoterSetOffset.orElse(null);
+                // if lastVoterOffset is -1 we know the leader hasn't written the bootstrap snapshot records to the log yet
+                if (voterSetOffset != null && voterSetOffset.offset() == -1) {
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/LeaderState.java]: Avoid the use of `null`.
```java
                lastVoterSetOffset.ifPresent(voterSetOffset -> {
                    // if lastVoterOffset is -1 we know the leader hasn't written the bootstrap snapshot records to the log yet
                    if (voterSetOffset.offset() == -1) {
```

File: raft/src/main/java/org/apache/kafka/raft/LeaderState.java [hunk 5]
```diff
@@ -202,7 +214,46 @@ public void appendLeaderChangeMessage(long currentTimeMs) {
...
+                    if (kraftVersion > 0) {
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/LeaderState.java]: It is an `IllegalStateException` if a voters set exist in the bootstrap checkpoint but the kraft.version is 0.

File: raft/src/main/java/org/apache/kafka/raft/internals/VoterSetOffset.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.raft.internals;
+
+public class VoterSetOffset {
+    private final VoterSet voterSet;
+    private final Long offset;
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/internals/VoterSetOffset.java]: An object is not needed. Use `long` instead. It takes less space and doesn't require a dereference.

File: raft/src/main/java/org/apache/kafka/raft/internals/VoterSetOffset.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.raft.internals;
+
+public class VoterSetOffset {
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/internals/VoterSetOffset.java]: Why did you add this type instead of using `LogHistory.Entry<VoterSet>`?

Comment (ahuang98) [reply raft/src/main/java/org/apache/kafka/raft/internals/VoterSetOffset.java]: I was on the fence about it, I thought this might be a bit more explicit/clear as a return type

Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/raft/internals/VoterSetOffset.java]: `LogHistory.Entry<VoterSet>` is explicit, no? It is a type that has an `offset()` and the `value()` of `VoterSet`.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+        RawSnapshotReader snapshot = context.log.latestSnapshot().get();
+        List<String> expectedBootstrapRecords = new ArrayList<>();
+        expectedBootstrapRecords.add(new SnapshotHeaderRecord()
+            .setVersion((short) 0)
+            .setLastContainedLogTimestamp(0).toString());
+        expectedBootstrapRecords.add(new KRaftVersionRecord()
+            .setVersion(ControlRecordUtils.KRAFT_VERSION_CURRENT_VERSION)
+            .setKRaftVersion((short) 1).toString());
+        expectedBootstrapRecords.add(voterSet.toVotersRecord(ControlRecordUtils.KRAFT_VOTERS_CURRENT_VERSION).toString());
+        expectedBootstrapRecords.add(new SnapshotFooterRecord().setVersion((short) 0).toString());
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: How about:
```java
        RawSnapshotReader snapshot = context.log.latestSnapshot().get();
        List<ApiMessage> expectedBootstrapRecords = Array.asList(
                new SnapshotHeaderRecord()
                    .setVersion((short) 0)
                    .setLastContainedLogTimestamp(0)),
                ...
        );
```

We should try to avoid comparing `String`. `ApiMessage` overrides `equals`.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+
+    @ParameterizedTest
+    @ValueSource(booleans = { true, false })
+    public void testLeaderWritesBootstrapRecords(boolean withKip853Rpc) throws Exception {
+        int localId = 0;
+        int follower = 1;
+        int observer = 2;
+        Uuid localDirectoryId = Uuid.randomUuid();
+        Uuid followerDirectoryId = Uuid.randomUuid();
+        Uuid observerDirectoryId = Uuid.randomUuid();
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: How about:
```java
ReplicaKey followerKey = replicaKey(1, withKip853Rpc);
```

You can see an example in `KafkaRaftClientTest`. I'll look at what you are testing but I am not sure that we should test `withKip853Rpc=false` since that is not theoretically possible if the kraft.version is 1.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+        RecordBatch batch = records.batches().iterator().next();
+        assertTrue(batch.isControlBatch());
+        Iterator<Record> recordIterator = batch.iterator();
+        Record record = recordIterator.next();
+        RaftClientTestContext.verifyLeaderChangeMessage(localId, Arrays.asList(localId, follower),
+            Arrays.asList(localId, follower), record.key(), record.value());
+        record = recordIterator.next();
+        RaftClientTestContext.verifyKRaftVersionRecord(kRaftVersion, record.key(), record.value());
+        record = recordIterator.next();
+        RaftClientTestContext.verifyVotersRecord(voterIds, record.key(), record.value());
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: Let's not increase the scope of `RaftClientTestContext` unless it is strictly necessary. Since these are `public static` method I doubt you need them in `RaftClientTestContext`. You can declare those methods as static in this type instead.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+            context.fetchRequest(
+                epoch,
+                ReplicaKey.of(follower, followerDirectoryId),
+                localLogEndOffset,
+                epoch,
+                0
+            )
+        );
+        context.pollUntilResponse();
+        context.assertSentFetchPartitionResponse(Errors.NONE, epoch, OptionalInt.of(localId));
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: This doesn't "check that leader sends records to follower" since you are fetching at the end of the log `context.log.endOffset().offset`.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+            KafkaRaftClientSnapshotTest.fetchSnapshotRequest(
+                context.metadataPartition,
+                epoch,
+                BOOTSTRAP_SNAPSHOT_ID,
+                Integer.MAX_VALUE,
+                0
+            )
+        );
+        context.pollUntilResponse();
+        context.assertSentFetchSnapshotResponse(Errors.NONE);
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: Hmm. Good point. The leader should not send this snapshot. Should the leader reply with:
```java
SNAPSHOT_NOT_FOUND(98, "Requested snapshot was not found.", SnapshotNotFoundException::new)
```

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+            ReplicaKey.of(leader, leaderDirectoryId)));
+        VoterSet voterSet = VoterSetTest.voterSet(voters.stream());
+        short kRaftVersion = 1;
+        int epoch = 5;
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(OptionalInt.of(localId), voterIds, localDirectoryId, kRaftVersion)
+            .withKip853Rpc(withKip853Rpc)
+            .withElectedLeader(epoch, leader)
+            .withBootstrapSnapshot(Optional.of(voterSet), kRaftVersion)
+            .appendToLog(epoch, Arrays.asList("a", "b", "c"))
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: Technically this is not a valid state. The whole point of your PR is that the start of the log should contain the control records (KRaftVersionRecord and VotersRecord) from bootstrap checkpoint before any other record.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+        VoterSet leadersVoterSet = VoterSetTest.voterSet(leadersVoters.stream());
+        List<String> leaderRecords = Arrays.asList(leadersVoterSet.toVotersRecord(ControlRecordUtils.KRAFT_VOTERS_CURRENT_VERSION).toString());
+        MemoryRecords batch = context.buildControlBatch(3L, epoch, leaderRecords);
+        context.deliverResponse(
+            fetchRequest.correlationId(),
+            fetchRequest.destination(),
+            context.fetchResponse(epoch, leader, batch, 3, Errors.NONE));
+
+        // this is broken because RaftClientTestContext.buildBatch does not send records in viable format
+        // follower applies the bootstrap records, registering follower2 as a new voter
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: Why is this broken? What do you mean by viable format?

Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: When the follower attempts to load the log with these records it throws an NPE because the record.key() is null

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java (hunk 13 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: this is testing the follower log - we wouldn't expect the control records (KRaftVersionRecord, VotersRecord), to be in the log until replicated from the leader correct?

File: raft/src/main/java/org/apache/kafka/raft/internals/VoterSetOffset.java (hunk 7 above)
Comment (ahuang98) [reply raft/src/main/java/org/apache/kafka/raft/internals/VoterSetOffset.java]: name-wise I thought it would be a bit more clear - `VoterSetOffset` vs `LogHistory.Entry<VoterSet>` but obviously that's a subjective opinion :) 

I'll remove this class though, it doesn't add any value functionally

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java (hunk 13 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: nevermind, I see why you're saying this is invalid - the follower replicates from the leader which should not have other records before these control records

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java (hunk 14 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: thanks for discussing this offline, I'll try appending control records differently

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java (hunk 12 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java]: do we want a more specific error message? (for now I've made the following KafkaRaftClient change)
```
         Optional<RawSnapshotReader> snapshotOpt = log.readSnapshot(snapshotId);
 -       if (!snapshotOpt.isPresent()) {
 +       // we do not allow fetching the bootstrap snapshot
 +       if (!snapshotOpt.isPresent() || snapshotId.equals(BOOTSTRAP_SNAPSHOT_ID)) {
            return RaftUtil.singletonFetchSnapshotResponse(
                ...
                    .setErrorCode(Errors.SNAPSHOT_NOT_FOUND.code())
```

**Comment by ahuang98** [comment]
@jsancio these are the open comments I haven't addressed yet - 
https://github.com/apache/kafka/pull/16518#discussion_r1669073620
https://github.com/apache/kafka/pull/16518#discussion_r1669183797 
https://github.com/apache/kafka/pull/16518#discussion_r1664929451
